### PR TITLE
Fix _dynamo.variables.torch_function.global_mangled_class_name

### PR DESCRIPTION
**Pull request by rec**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132744

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Comment by rec** [comment]
The comment on the code says:

```
        # The global_mangled_class_name should be different for different
        # invocations of torch.compile. Otherwise, we can run into a situation
        # where multiple torch.compile invocations re-use the same global name,
        # but the global's lifetime is tied to the first invocation (and
        # may be deleted when the first torch.compile invocation is deleted)
        # We mangle it based off of the output_graph's id.
```

...but instead of using the compile_id, it mangles the name with the built-in _function_ `id`, which is always the same, probably due to a typo.

----

The local variable `compile_id` was assigned and not used. 

`pytorch` installs `pylint`, which would have caught this, but the `lintrunner` machinery doesn't appear to use `pylint` at this time.

I see that `pylint` also flags many conditions in the existing code which probably should not be checked, but it would appear to be fairly easy to add `pylint` to `lintrunner` and only have it check for unused locals, which are almost always an error.

**Review by amjames — COMMENTED** [review]
Fix looks good. 

Some kind of test should be added (or updated if it already exists) 

I am curious though: this was clearly not behaving correctly for a while, so is there a gap in testing or is it even used/needed?

**Comment by rec** [comment]
Oh, it's definitely used: it gets called when you subclass `Tensor` and then compile, through [this call](https://github.com/pytorch/pytorch/blob/8b50d5398fab79b599178627bed766859b3cd9dd/torch/_dynamo/variables/torch_function.py#L207-L210).

A code example is [here](https://github.com/pytorch/pytorch/blob/8b50d5398fab79b599178627bed766859b3cd9dd/test/dynamo/test_modules.py#L1269-L1284) and I was tracing through that test when I ran into this (extraneous) issue.

There are also two other spots in the code where this method appears to be called, one also in [torch_function.py](https://github.com/pytorch/pytorch/blob/8b50d5398fab79b599178627bed766859b3cd9dd/torch/_dynamo/variables/torch_function.py#L198-L202), and one in [codegen.py](https://github.com/pytorch/pytorch/blob/8b50d5398fab79b599178627bed766859b3cd9dd/torch/_dynamo/codegen.py#L167), but I haven't run into them yet.

I kind of feared you were going ask for a unit test :-D even though it's very logical because I'm not sure exactly how to do this yet (in a non-stupid way, I could always construct a `TensorWithTFOverrideVariable` by hand and just call the function). I'm hoping once I resolve the issue I was debugging above I'll use my knowledge to create this test - stay tuned.

**Comment by rec** [comment]
I did in fact come up with a useful test that shows that we generate a different mangled class name for each compile_id.

I added it into `test/dynamo/test_modules.py` because it shares test code with two other classes in this file, but I'd easily be persuaded to move it elsewhere - I couldn't figure out where the right spot would be.

File: test/dynamo/test_modules.py [hunk 1]
```diff
@@ -1330,11 +1333,61 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
...
+
+                    class_type_id, compile_id = after.split("_")
+                    self.assertTrue(class_type_id.isnumeric())
+                    self.assertTrue(compile_id.startswith("c"))
+
+                    cid = compile_id[1:]
+                    self.assertTrue(cid.isnumeric())
+                    compile_ids.add(int(cid))
+
+                self.assertEqual(compile_ids, set(range(3)))
```
Comment (zou3519) [inline test/dynamo/test_modules.py:1386]: We should just check that they're unique, not the exact value. The exact value might be flaky

**Review by zou3519 — APPROVED** [review]

File: test/dynamo/test_modules.py (hunk 1 above)
Comment (rec) [reply test/dynamo/test_modules.py:1386]: Very good point, I just collected the compile ids as strings into a set, and checked that the size of the set was 3.

Comment (rec) [reply test/dynamo/test_modules.py]: Hah, right after you said this, one instance of this unit test failed, showing that this wasn't hypothetical. :-D Let's see how the CI shakes out, but the last change did seem to fix that test locally.

**Pull request closed without merging**